walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -18,18 +18,138 @@
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
-import java.util.Map;
 import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * Return the TimeRanges available for all meeting members given existing
+   * events
+   */
+  public Collection<TimeRange> query(Collection<Event> events,
+                                     MeetingRequest request) {
+    // For a meeting longer than a day, no times are possible
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // Get collections of people from MeetingRequest
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // With no attendees, the meeting can be held at any time
+    if (attendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    long duration = request.getDuration();
+
+    // If either group is empty, only need to consider the other
+    if (attendees.isEmpty() || optionalAttendees.isEmpty()) {
+      Collection<String> nonEmptyAttendees =
+          attendees.isEmpty() ? optionalAttendees : attendees;
+      return combineAvailability(events, nonEmptyAttendees, duration);
+    }
+
+    ArrayList<TimeRange> mandatoryAvailability =
+        combineAvailability(events, attendees, duration);
+
+    // If there are no open slots between mandatory attendees,
+    // no need to consider optional attendees
+    if (mandatoryAvailability.isEmpty()) {
+      return mandatoryAvailability;
+    }
+
+    // Get individual availabilities for optional attendees
+    ArrayList<ArrayList<TimeRange>> optionalAttendeesCals =
+        new ArrayList<ArrayList<TimeRange>>();
+        
+    for (String curPerson : optionalAttendees) {
+      optionalAttendeesCals.add(
+          combineAvailability(events, Arrays.asList(curPerson), duration));
+    }
+
+    // Meetings can be held every {increment} minutes
+    // E.g. if all meetings are only held at 15 minute increments of the hour
+    // for durations that are multiples of 15 minutes, can reduce amount of
+    // computation by setting {increment} to 15
+    int increment = 1;
+
+    // For all possible times every {increment} minutes that allows for a 
+    // meeting of the desired duration, store the number of optional attendees 
+    // who can attend
+    ArrayList<ArrayList<Integer>> forEachDuration =
+        new ArrayList<ArrayList<Integer>>();
+    
+    for (int i = 0; i < mandatoryAvailability.size(); i++) {
+      forEachDuration.add(new ArrayList<Integer>());
+      int n = 0;
+      while (n * increment + duration <=
+             mandatoryAvailability.get(i).duration()) {
```

Comment:
I'm not following why you are increasing the duration by n*increment?

Don't you want to check if the meeting starts at 0:1440 and lasts the duration how many people can go?

Reply:
This shifts the duration by `increment` minutes until the end of the slot in the mandatory attendees' calendar.
```
|--|    increment
|-----| duration

|-----|          Duration-length slot to check availability of all optional attendees, n = 0
|-------------|  Time slot where all mandatory attendees are available

|--|-----|       Duration-length slot to check availability of all optional attendees, n = 1
|-------------|  Time slot where all mandatory attendees are available
.
.
.
|--|--|--|-----| n++ until no more duration-length slots possible
|-------------|  Time slot where all mandatory attendees are available
```
If I understand what you're saying about 0:1440, I have not accounted for midnight represented as either 0 or 1440? The TimeRange class defines a day as going from 0 to 1439 minutes, and my code works for a meeting potentially starting at 0 minutes.

Reply:
Oh, I wasn't asking about midnight that was just shorthand for the range [0,1,2,3,....,1440].  

Ah, now I understand, I thought you were changing the duration.

I think this would be more clear if you changed this to a for loop and pulled out the incrementedStart that I commented out below.  So something like:

`
// Iterate over all the potential meeting blocks of size duration and count the number of optional attendees that are free.

for (int potentialStartTime = mandatoryAvailability.get(i).start(); potentialStartTime < mandatoryAvailability.get(i).end() - duration; potentialStartTime++) {
   ...
}
`